### ISSUE TITLE
Problems with analyses cache headers and invalidations

### ISSUE THREAD

**Issue by rochoa** [CartoDB/Windshaft-cartodb#629]
## Prerequisite

1. Use Builder to simplify the STR.
1. Have a points dataset `s1`.

## STR

1. Create a Map with our `s1` points dataset.
1. Add a new analysis: "Create areas of influence", use default params.
   - This creates a temporary table for the analysis, let's say `b1`.
1. Go to the embed and check the headers for the Maps API Named Map instantiation:
   - X-Cache-Channel: will be `b1`, from the analysis table.
   - Surrogate-Key: will contain the named map key, `n:{hash(name)}`; but also the table key: `t:{hash(b1)}`.
1. Run, to simulate the dataset changed, the query:
```sql
select cdb_tablemetadatatouch('s1');
```
5. Reload the named map instantiation and check the headers again:
   - X-Cache-Channel: will be `b2`, as the source table changed, the analysis table is new and has a different name.
   - Surrogate-Key: the same happens with the table key: `t:{hash(b2)}`.

## Problems with this approach

This, per se, it's not a problem because the headers have the correct headers.

However, this becomes a problem when there is a cache layer relying on those headers for its invalidation mechanism. Actually, there are several problems:

1. The named map response never contains any header associated with the original `s1` table. Once that table changes (in our scenario due to `cdb_tablemetadatatouch`), we don't invalidate the response because it was never tag/annotated with any information about the original `s1` table.
1. As analyses tables are not cartodbifyed, they won't trigger any invalidation. In our case that means that once `s1` is modified and the analysis runs again, we won't invalidate any object tagged with `b1` automatically. As an alternative, we could manually trigger an invalidation for `b1` once the analysis is computed again, but at that point, the analysis is totally different because `s1` already changed, and we will only know about `b2` and nothing about `b1` anymore.
1. Another alternative is to tag the named map with the original `s1` table name. That way, we will invalidate the named map when the data at `s1` changes. However, that introduces a race condition, because if the named map is instantiated again before the analysis for `b2` finishes, we will cache it again before the analysis finished.

## Potential solution

1. Tag the named map response with the original `s1` table name.
1. Each analysis invalidates all objects associated to the source tables they rely on after the analysis is finished, avoiding the race condition.

The problem with this approach is we will invalidate other objects associated to the original `s1` table, and not only the named maps associated to analyses relying on it.

**Comment by rochoa** [comment]
Related issue: https://github.com/CartoDB/cartodb/issues/11157.

**Comment by ethervoid** [comment]
After talking with @rochoa:

#### Source table's cached response invalidation

We need to invalidate correctly the source table's cached response.

Right now the `Surrogate Key` header contains:
    - `namespace` key composed of the named map template id
    - `table` key composed of the last analysis table, if any, or the source table if there isn't any analysis.

The broken case happens when the map has analysis, with associated tables, and the source table changes. In that case, we try to invalidate
the cache using the `source table` name but we have tagged the cached responses with the `analysis table` name so is not possible to invalidate
it.

The proposed solution is to add a new surrogate key using the `source table` name.

#### Race condition between invalidate source table and analysis generation

This is a tricky one. Let's put it through an example.

We a have a table `t1` and an analysis `a1`. We make a change in the analysis parameters and it starts running again. In the meantime, we make a change in the table `t1` data which leads to a cache invalidation (we're assuming the prior bug is fixed).

Now we enter in the `embed` map. This won't use the cached response because is invalidated, so it makes a request and we cache the response while the analysis `a1` is still running. And here it is the issue.

Now we have the response cached with the analysis `a1` in the `running` state and when it finish it won't be invalidated.

The proposed solution for this issue is to invalidate the cache, using the source table key, when any analysis finish

**Comment by ethervoid** [comment]
## Solving the cache problem

#### Tasks

- [ ] Add the source table name as surrogate key in order to be able to invalidate it when the table changes
- [ ] Reproduce the race condition
- [ ] Make the analysis invalidate the cache when finish (using the source table name as key)
- [ ] Tests for this part

#### Get the source table for a layer

We need to retrieve the source table name from a layer and we have two cases:

- If we don't have analysis, we take the affected tables from the layer, which is going to be the source table
- If that layer has, at least, one analysis, we need to get the source table/s from the source/s analysis.

#### Affected tables with layers SQL

We get the affected tables throught the layer's query. The not cool part of this is that the layer uses the last generated table.

For example:

- We create an empty map. The layer query is "select * from untitled_map_1".
- Now we add some data to that table and also we add an analysis with cached table
- Now the layer query is "select * from analysis_{hash}" and we aren't able to get the source table because is shadowed by the analysis table

In summary, we can get the source table using affected tables with the layers

#### Analysis config

This seems to be the way to go, even an empty "map" has analysis with a source node, which is what we want, so if we're able to get the analysis config we could make a call to camshaft and parse it to get the source node.

The problem is that this config seems to be passed as parameters and we don't store it.

**Comment by rochoa** [comment]
Quick reminder, we should take the same approach for both, the `surrogate-keys` and the `x-cache-channel` headers.

**Comment by ethervoid** [comment]
Yes, once we have the source table we could use it to add it to the Surrogate keys and to the X-Cache-Channel headers.

Right now my approach is as follows:

- Add logic to get the analysis config from mapconfig in the `AnalysisBackend` using the `map_cdg|token_id` or `map_tpl|username` in the named maps case (ie. NamedMapController)
- Parse this mapconfig to extract the source node's table name. We could do this in two ways:
    - Make camshaft aware of how to parse analyses part from map config, which I don't think is the best way becuase `camshaft` should be agnostic of this part
    - Make the Analysis backend to extract the source node's table name
    - Extract the parsing functionallity from `AnalysisMapConfigAdapter` and use it to get the source node using `camshaft`. I think this is the best way to do it
- One we have the source table name, we could use it to create the key used in `Surrogate keys` and `X-Cache-Channel``

What do you think @rochoa?

**Comment by ethervoid** [comment]
I'd include new functions in the [node-cartodb-redis](https://github.com/CartoDB/node-cartodb-redis) repository to get the map_tpl and/od map_cfg data

**Comment by rochoa** [comment]
What about changing camshaft's `Node` to keep track of the affected tables? We are already computing the `updatedAt` for Source nodes from the query using `CDB_QueryTables_Updated_At`.

**Comment by ethervoid** [comment]
After talking offline with @rochoa:

- We're going to add the affected tables as a new property in the nodes. That way we could ask for the affected tables 
- Use the *Adapter to include this new field to have it when needed

**Comment by ethervoid** [comment]
## Acceptance

We need to check two cases in here:

- Invalidation is called against the source table/s instead against the "analysis" table
- When an analysis finish, we've to invalidate all the related cache for the whole analysis. To do it we invalidate them using the source table as cache tag.

These are the acceptance steps (should be done for an org and non-org user account):


1. Import this dataset [invalidation.csv.zip](https://github.com/CartoDB/Windshaft-cartodb/files/844110/invalidation.csv.zip) and activate the labels with the `testdate` column. You have to convert it from date to string be able to use that column as a label.
2. Open the Chrome dev tools -> Network and filter by `stat_tag` in order to get the map instantiation request.
3. Open the map created with the provided dataset or refresh the screen. 
4. Check the headers and you should have:
    - In the X-Cache-Channel header something like this: `cartodb_staging_user_b19a3b93-9826-4cf1-8c69-60d8871a31cc_db:"ethervoid-rui01-owner".invalidation`
    - In Surrogate-key something like this: t:O1URni
5. Execute this query: https://{USERNAME}.carto-staging.com/api/v2/sql?q=update%20invalidation%20set%20testdate%20=%20now()::text;&api_key={API_KEY}`
6. Refresh the map and check the labels show the new date
7. Remove the `stag_tag` filter in Chrome dev tools and check other tiler requests, for example the tile request and check the headers too
8. Create an analysis, for example, Areas of interest
9. Check the headers again, now you should see something like this:
    - In the `X-Cache-Channel` header something like this: `cartodb_staging_user_b19a3b93-9826-4cf1-8c69-60d8871a31cc_db:"ethervoid-rui01-owner".invalidation,"ethervoid-rui01-owner".analysis_112d4fc091_2be49579ee543ae204179bf2d663b03430f8499e`
    - In `Surrogate-key` something like this: t:O1URni t:Xy6wjz
10. Execute the query again and refresh. The labels should show now the new date
11. Now to check the invalidation when the analysis is updated, for example changing the parameters, you need to check the `TIS Invalidations` dashboard in Staging's Kibana (ask me about it)
12. Change the parameters and check Kibana for an entry like this `TCH,cartodb_staging_user_b19a3b93-9826-4cf1-8c69-60d8871a31cc_db,"ethervoid-rui01-owner".invalidation` in the `command` field. This means that TIS is invalidating all the caches related to that table which is the source one.
13. Remove the `stag_tag` filter in Chrome dev tools and check other tiler requests, for example the tile request and check the headers too

**Comment by dgaubert** [comment]
I've followed all steps of the acceptance and everything works fine for a non-org user. I have a silly question about the header for a tile response. Does tile response have a `Surrogate-Key`header?. If so, I can't see it:

![screen shot 2017-04-05 at 15 42 19](https://cloud.githubusercontent.com/assets/1936888/24708222/8682d080-1a16-11e7-9edd-66b22b94037d.png)

I'll do same steps for an org-user later.

**Comment by ethervoid** [comment]
Yes, it should have the `Surrogate-Key` header: https://github.com/CartoDB/Windshaft-cartodb/blob/master/lib/cartodb/controllers/layergroup.js#L370

**Comment by dgaubert** [comment]
So I guess we should take a look, shouldn't we?

**Comment by ethervoid** [comment]
Yep, tomorrow I'll take a look at it. Thank you :)

**Comment by ethervoid** [comment]
I've checked in a server with fix and without fix and the behavior is the same, we have `X-Cache-Channel` but we don't have `Surrogate-Key` this seems like a bug we haven't spotted yet.

Another thing I've seen is that we don't have these headers in the production tiles:

![captura de pantalla 2017-04-05 18 38 10](https://cloud.githubusercontent.com/assets/741240/24716344/0c298b12-1a2f-11e7-896d-e30b1d868f9a.png)

**Comment by dgaubert** [comment]
I've just followed steps with a user in a organization and it has the same behaviour. By the way, I realized that if I change an analysis param, it looses the table value for `X-Cache-Channel` header. For example:

AOI:
![screen shot 2017-04-06 at 12 42 20](https://cloud.githubusercontent.com/assets/1936888/24750754/506c0690-1ac7-11e7-9a7b-7ae14846abf6.png)

Changing radius:
![screen shot 2017-04-06 at 12 44 03](https://cloud.githubusercontent.com/assets/1936888/24750785/703df23a-1ac7-11e7-96cf-4bf2707c99df.png)

No sure if it is a bug or the right way.

Hope this helps

**Comment by ethervoid** [comment]
Hum looks like a bug because it should keep the affected tables either in `X-Cache-Channel` and in `Surrogate-Key` :S

**Comment by dgaubert** [comment]
I'm debugging a bit further and if I change the style the soruce table value for `X-Cache-Channel` is back:

![screen shot 2017-04-06 at 13 13 23](https://cloud.githubusercontent.com/assets/1936888/24751812/3ec4a1e6-1acb-11e7-9456-e0027476c6e0.png)

**Comment by ethervoid** [comment]
Yesterday we were debugging and the root cause was that we have multiple APIs in the staging cloud and one of them was using Node6 and it hasn't this fix deployed.

**Comment by dgaubert** [comment]
BTW I'd check if tile responses need 'Surrogate-Key' header. I don't know if we are missing them or it is ok.

**Comment by ethervoid** [comment]
Yep, I'm on it \o/

**Comment by ethervoid** [comment]
I've checked it and I've found why is working that way:

- Local environment: We filter neither `Surrogate-Key` nor `X-Cache-Channel`
- Staging environment: Here we filter `Surrogate-Key` [here](https://github.com/CartoDB/cartodb-platform/blob/087145b031c91e4277580068f7a5cd2c8fa9d8d6/deploy/fastly/staging.vcl#L294-L297) but we don't filter `X-Cache-Channel` as you can see [here](https://github.com/CartoDB/cartodb-platform/blob/1183b4b5fd2846ef32cc1a57f3650d236358f605/cookbooks/cartodb_app/templates/default/vhost.cartodb.com.erb#L124-L133)
- Production environment: We filter either `Surrogate-Key` ([here](https://github.com/CartoDB/cartodb-platform/blob/087145b031c91e4277580068f7a5cd2c8fa9d8d6/deploy/fastly/remove_varnish_headers.vcl#L16-L19)) and `X-Cache-Channel`([here]())

So, problem solve. The Headers are being generated, as you can see in your local environment, but have different behaviors in staging and production. 

Maybe we should let the `Surrogate-Key` header unfiltered in staging. What do you think @luisbosque or @azamorano

**Comment by dgaubert** [comment]
Nice!

Thanks for debugging the root cause!

**Comment by ethervoid** [comment]
I've made some changes in the camshaft project:

- [Perform an extraction](https://github.com/CartoDB/camshaft/pull/269/commits/bfa1731f9ac8bde6e7f6615fe840005bccd12017) in one of the new functions
- Renamed [this function](https://github.com/CartoDB/camshaft/pull/269/commits/a21c3ccd0cb3d42aaa5f18631d701dc99ca60e98) to make it more readable

also, I've rebased the `windshaft-cartodb` part against `master` so we should be ready to deploy it for the 3.x servers.

I've made a [new branch](https://github.com/CartoDB/Windshaft-cartodb/tree/fix_invalidation_for_v2x_version) based on `v2.x` branch with all the changed cherry-picked.

**Comment by ethervoid** [comment]
Deployed, thank you very much to @dgaubert and @rochoa for all your help in my first serious issue :)
